modules/DGM/PSO.cpp

```diff
@@ -0,0 +1,215 @@
+//
+// Created by ahambasan on 22.02.20.
+//
+
+#include "PSO.h"
+#include "macroses.h"
+
+#include <iterator>
+#include <vector>
+#include <random>
+
+DirectGraphicalModels::PSO::PSO() = default;
+
+DirectGraphicalModels::PSO::PSO(const vec_float_t &vParams)
+        : m_nParams(vParams.size()),
+          m_vParams(vParams.size()),
+          m_vMin(vParams.size()),
+          m_vMax(vParams.size()),
+          isThreadsEnabled(false) {
+    for (size_t p = 0; p < vParams.size(); p++) {
+        const float &param = vParams[p];
+        m_vParams[p] = param;
+    }
+    m_nParams = vParams.size();
+
+    // SOURCE: https://en.cppreference.com/w/cpp/numeric/random/uniform_real_distribution
+    std::random_device rd;  //Will be used to obtain a seed for the random number engine
+    std::mt19937 gen(rd());
+    std::uniform_real_distribution<> dis(-10.00f, 10.00f);
+
+    // initialize boids
+    for (auto i = 0; i < NUMBER_BOIDS; i++) {
+        Boid *b_x = new Boid;
+        for (auto d = 0; d < m_nParams; d++) {
+            b_x->pParams.push_back(dis(gen));
+            b_x->velocity.push_back(1);
+        }
+        b_x->pBest = m_vParams;
+        b_n.push_back(*b_x);
+    }
+
+    gBest = m_vParams;
+
+    // initialize meta parameters
+    c1 = C1_DEFAULT_VALUE;
+    c2 = C2_DEFAULT_VALUE;
+    w = W_DEFAULT_VALUE;
+
+    reset();
+}
+
+void DirectGraphicalModels::PSO::reset() {
+    std::fill(m_vMin.begin(), m_vMin.end(), -FLT_MAX);
+    std::fill(m_vMax.begin(), m_vMax.end(), FLT_MAX);
+    std::fill(m_vParams.begin(), m_vParams.end(), 0.0f);
+}
+
+void DirectGraphicalModels::PSO::setInitParams(const vec_float_t &vParams) {
```

Comment:
I did not went deep into the algorithm, but maybe the params may be shared also with Powell? If yes, maybe move this function to the base class?

Reply:
The base class is an abstract base class (more of an interface).

Reply:
I think I saw some function with bodies in the base abstract class. So it is not an _interface_ class so to say. I think we can move the common code from both Powell and PSO there...